### PR TITLE
feat(pacer): Link pacer sessions to proxies

### LINKED ISSUE #4087: Update catching strategy to link session data to proxy connections
Following PACER's confirmation that we should use a static IP address to crawl pages. I've reviewed the code and believe we can link cookies with proxy connections.  The `log_into_pacer` method is the key for this refactor.

we'll need to refactor the code in the following ways:

- Add a new argument to the [ProxyPacerSession](https://github.com/freelawproject/courtlistener/blob/7aac7069eecd15d6b9a1ec0b092fdf73db8e2561/cl/lib/pacer_session.py#L15-L71) class constructor:

   The current `ProxyPacerSession` class constructor retrieves the proxy information from the settings. This limits the flexibility of the class, as it cannot be used with requests that require a different proxy than the one configured in settings.

    We can achieve greater flexibility by adding a new optional proxy argument to the constructor. This enhancement ensures that all traffic associated with a specific session is directed through the same proxy (we’ll use the same proxy used to create the cookie).

- Override the login method in the `ProxyPacerSession`:

   This will allow us to achieve more flexibility in proxy selection, if no proxy is provided during initialization, the overridden login method will handle selecting a suitable proxy before proceeding with the login process. Here's a code snippet demonstrating the override:

```python
def login(self):
        if not self.proxy:
           # Implement logic to select a suitable proxy
           self.proxy = logic_to_select_a_proxy()
        return super().login()
```

- Tweak the [log_into_pacer](https://github.com/freelawproject/courtlistener/blob/7aac7069eecd15d6b9a1ec0b092fdf73db8e2561/cl/lib/pacer_session.py#L74-L92) method:
   
   - Instead of just returning the cookie, we should return both the cookie and the selected proxy. This could be a dictionary or a tuple to hold both values.

- Store the `cookie-proxy` pair and update caching strategy

   Our caching strategy needs an update to accommodate the new cookie-proxy pairing. We currently retrieve the cached cookie by simply checking for a key and unpickling it. However, with the refactoring, the cached data could be either:

   1. A cookie-proxy pair represents the desired outcome, with the login cookie and the associated proxy information.
   2. Just the cookie (existing approach)

    For cached entries containing only the cookie (scenario 2), we can use the flexibility of proxies. Here's the approach:

    1. We'll randomly choose a proxy from the available pool.
    2. We'll attempt to complete tasks using this newly selected proxy and the existing cookie.
    3. If a PacerLoginException occurs, it might indicate an issue with the chosen proxy.
    4. Since we retry those exceptions, the retrieval logic will have a chance to select a different proxy during the next attempt.
    5. This way, we can handle existing cookies in the cache while transitioning to the new cookie-proxy pair approach.

### PR DESCRIPTION
**Pull request by ERosendo**
This PR introduces a new approach to caching PACER sessions and updates the code of the ProxyPacerSession class to pick one of the proxy connection strings from the settings.

Key Changes:

- Introduces a new environment variable (EGRESS_PROXY_HOSTS) to manage a list of proxies.

- Implements new logic in the ProxyPacerSession  to choose a proxy connection string from the settings, enabling dynamic proxy selection.

- Updates the `log_into_pacer` method to return a tuple containing both the user's cookie and the used proxy address used for login

- Tweaks the `get_or_cache_pacer_cookies` to handle both the new return format and potentially existing cached data in the old format, ensuring backward compatibility.

- Updates tasks from `corpus_importer` and `recap` modules to accommodate the new PACER user cookie format.

- Adapts commands in the `corpus_importer` module that use cookies to handle the new format.

- Introduces new tests for pacer session utils.

This PR closes #4087

### CODE REVIEW

File: cl/corpus_importer/tasks.py [hunk 1]
```diff
@@ -988,10 +990,21 @@ def get_pacer_case_id_and_title(
...
+        cookies, proxy_address = cookies_data
+    elif user_pk:
+        cookies_from_cache = get_pacer_cookie_from_cache(user_pk)
+        if isinstance(cookies_from_cache, tuple):
+            cookies, proxy_address = cookies_from_cache
+        cookies, proxy_address = cookies_from_cache, settings.EGRESS_PROXY_HOST
+    else:
+        # If the user_pk is unavailable, cookies cannot be retrieved from the
+        # cache.
+        self.request.chain = None
```
Comment (mlissner) [inline cl/corpus_importer/tasks.py:1004]: Is it better to throw an exception here? In theory we should never arrive at this code, right?

**Review by mlissner — CHANGES_REQUESTED** [review]
Looks like you've got an effective solution here, and that you've thought carefully about how to roll it out without queued tasks failing. Nice. 

I think I would change something though. Having to work with tuples and pass them around is never much fun. They're kind of ugly, it's hard to remember what order the tuple is in ("Was it cookies then proxy or proxy then cookies?"), and so forth. 

Since I think the values we need are getting added to the PACER session object anyway, did you consider just passing that around? Even `log_into_pacer` could just return `s`, I think?

File: cl/corpus_importer/tasks.py (hunk 1 above)
Comment (ERosendo) [reply cl/corpus_importer/tasks.py:1004]: > In theory we should never arrive at this code, right?

You're right! I'll update the code to throw an exception.

File: cl/recap/tasks.py [hunk 2]
```diff
@@ -1656,8 +1661,13 @@ def fetch_attachment_page(self: Task, fq_pk: int) -> None:
         mark_fq_status(fq, msg, PROCESSING_STATUS.FAILED)
         return
 
+    cookies_data = (
+        cookies
+        if isinstance(cookies, tuple)
+        else (cookies, settings.EGRESS_PROXY_HOST)
+    )
```
[the remark this answers is not in the document]
Comment (ERosendo) [reply cl/recap/tasks.py:1668]: I'll add a comment! 👍

**Comment by ERosendo** [comment]
@mlissner  I've implemented your suggestions. I've removed the `EGRESS_PROXY_HOST` environment variable from settings and replaced tuples with a dedicated dataclass named `SessionData` for better structure and readability.

Let me know what you think.

**Review by mlissner — APPROVED** [review]
Great! This is a lot easier to reason about. Thanks for making the little helper class.

Can you make an issue with details about deployment and the follow up tasks? E.g.:

 - Need the variables set before deployment
 - Need to remove the old proxies after a few hours or maybe a day
 - Need to remove shim code after a few hours or maybe a day

Anything else?

**Comment by mlissner** [comment]
This is approved, but can you work with Ramiro to make sure that we've got the variables and k8s deployments in place before we merge and deploy?

**Comment by ERosendo** [comment]
Sure!

**Comment by ERosendo** [comment]
@mlissner Ramiro confirmed the environment variable was set up correctly and that CL can successfully reach the newly deployed proxies.

Let's merge this PR. :shipit:

**Comment by mlissner** [comment]
Auto-merge is set. Thanks all!

**Comment by mlissner** [comment]
@ERosendo Now that this has been live for a couple weeks, is there code in here we want to refactor to simplify things?

**Comment by ERosendo** [comment]
@mlissner Yes, We used several ternary if statements to manage legacy cached data. I believe we can safely remove this code. Here are examples:

https://github.com/freelawproject/courtlistener/blob/57162841fd5aeda54bab59ee0b8439652fcda6e6/cl/recap/tasks.py#L1658-L1660

https://github.com/freelawproject/courtlistener/blob/57162841fd5aeda54bab59ee0b8439652fcda6e6/cl/recap/tasks.py#L2416-L2421

https://github.com/freelawproject/courtlistener/blob/57162841fd5aeda54bab59ee0b8439652fcda6e6/cl/lib/pacer_session.py#L165-L169

**Pull request merged**